### PR TITLE
400-error

### PR DESCRIPTION
**Pull request by bean00**
I'm not satisfied with my `parseRequest` method in `RequestParser`. I'd like to see if I can clean up the method somehow, but I'm not sure how to do so at the moment. I'm not sure if extracting out part of it into a separate class might help.

Also, right now I'm assuming that the client only sends requests for the HTTP `1.1` protocol (I'm hardcoding the `HTTP/1.1` part of the request as a String). If requested, I will add in the code to parse the last part and check the version number.

### CODE REVIEW

File: src/main/java/com/bean00/Method.java [hunk 1]
```diff
@@ -1,7 +1,24 @@
 package com.bean00;
 
+import java.util.HashSet;
+
 public class Method {
     public static final String GET = "GET";
     public static final String HEAD = "HEAD";
 
+    private static HashSet<String> requestMethods = buildMethodsSet();
```
Comment (bean00) [inline src/main/java/com/bean00/Method.java]: Don't need the `requestMethods` set for now (might need later for Method Not Allowed story)

File: src/main/java/com/bean00/Method.java [hunk 2]
```diff
@@ -1,7 +1,24 @@
...
+
+    public static boolean isMethodValid(String method) {
+        return requestMethods.contains(method);
+    }
+
+    private static HashSet<String> buildMethodsSet() {
```
Comment (bean00) [inline src/main/java/com/bean00/Method.java]: Use `Set` interface instead of `HashSet`

File: src/main/java/com/bean00/RequestParser.java [hunk 3]
```diff
@@ -7,18 +7,40 @@
...
+            throw new BadRequestHttpException();
+        }
+        if (requestLine == null) { throw new BadRequestHttpException(); }
+
+        requestLine = requestLine.trim();
+        String[] requestWords = requestLine.split("\\s+");
+        if (requestWords.length != EXPECTED_NUMBER_OF_REQUEST_WORDS) { throw new BadRequestHttpException(); }
 
         String requestMethod = requestWords[0];
+        if (!Method.isMethodValid(requestMethod)) { throw new BadRequestHttpException(); }
```
Comment (bean00) [inline src/main/java/com/bean00/RequestParser.java]: Shouldn't check if the method is valid here. Put off for now until the Method Not Allowed story

File: src/main/java/com/bean00/Method.java [hunk 4]
```diff
@@ -1,7 +1,24 @@
...
 
+import java.util.HashSet;
+
 public class Method {
     public static final String GET = "GET";
     public static final String HEAD = "HEAD";
 
+    private static HashSet<String> requestMethods = buildMethodsSet();
+
+    public static boolean isMethodValid(String method) {
```
Comment (bean00) [inline src/main/java/com/bean00/Method.java]: Don't need for now

File: src/main/java/com/bean00/RequestParser.java [hunk 5]
```diff
@@ -7,18 +7,40 @@
...
 
-    public Request parseRequest() throws IOException {
-        String requestLine = in.readLine();
-        String[] requestWords = requestLine.split("\\s");
+    public Request parseRequest() {
+        String requestLine;
+        try {
+            requestLine = in.readLine();
+        } catch (IOException e) {
+            throw new BadRequestHttpException();
```
Comment (bean00) [inline src/main/java/com/bean00/RequestParser.java]: Maybe don't throw the `BadRequestHttpException` here (let the `IOException` fall through)

File: src/main/java/com/bean00/RequestParser.java [hunk 6]
```diff
@@ -7,18 +7,40 @@
...
+        }
+        if (requestLine == null) { throw new BadRequestHttpException(); }
+
+        requestLine = requestLine.trim();
+        String[] requestWords = requestLine.split("\\s+");
+        if (requestWords.length != EXPECTED_NUMBER_OF_REQUEST_WORDS) { throw new BadRequestHttpException(); }
```
Comment (bean00) [inline src/main/java/com/bean00/RequestParser.java:23]: Can try to use Java Optional here (can look into the Vavr Optional as well)

File: src/test/java/com/bean00/MethodTest.java [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class MethodTest {
+
+    @Test
+    public void isMethodValid_returnsFalse_ifTheMethodIsNotValid() {
```
Comment (bean00) [inline src/test/java/com/bean00/MethodTest.java]: Don't need these anymore

Comment (bean00) [reply src/test/java/com/bean00/MethodTest.java]: Removed

File: src/main/java/com/bean00/RequestParser.java (hunk 6 above)
Comment (bean00) [reply src/main/java/com/bean00/RequestParser.java:23]: Will hold off for now. Might try again later

File: src/main/java/com/bean00/Method.java (hunk 4 above)
Comment (bean00) [reply src/main/java/com/bean00/Method.java]: Removed

File: src/main/java/com/bean00/Method.java (hunk 1 above)
Comment (bean00) [reply src/main/java/com/bean00/Method.java]: Removed

**Review by jdesrosiers — APPROVED** [review]
It would be nice if you could find a way to remove the duplication (throwing BadRequest in three different places), but it's OK for now.

**Review by tbaik — APPROVED** [review]
Looks good! Same thing as Jason. I was hoping there can be something done about throwing it 3 times but I guess you need the checks

**Pull request merged**